### PR TITLE
Normalizing flow tutorial

### PR DESCRIPTION
**Pull request by ncfrey**
This PR includes a tutorial notebook for the normalizing flow API.

It covers the following:
- Taking the QM9 dataset of small organic molecules, translating them into SELFIES strings and then into vector representations 
- Training a simple normalizing flow on the target distribution of QM9 molecules
- Generating new molecules and translating them back into SELFIES and SMILES strings
- Comparing some generated molecules to QM9 through Tanimoto similarity

TODO:
- Update the `selfies` import when the new release is available via pip
- Update the tutorial number and tutorial README following the reorganization in #2112 

CC @rbharath @seyonechithrananda if you get a chance to check out the tutorial, I'd appreciate any feedback!

### CODE REVIEW

**Comment by peastman** [comment]
Looks good!  I have a few suggestions.

A lot of readers probably aren't familiar with SELFIES.  It would be helpful to add a few sentences discussing their advantages over SMILES.  You also could show the two representations of a few molecules to help people understand the differences.

Rather than downloading files from AWS and loading them with pandas, why not just call `dc.molnet.load_qm9()`?

I don't understand what the permutation layer is for.  It seems to just swap the first and second halves of the array.  Can you add more explanation of that?

This code

```python
for epoch in range(max_epochs):  # max_epochs
    loss = nfm.fit(ds, nb_epoch=1)
    losses.append(loss)
```

can be simplified to just

```python
loss = nfm.fit(ds, nb_epoch=1, all_losses=all_losses)
```

A few points about the following code:

```python
max_epochs = int(1e2)  # maximum number of epochs of the training
opt = Adam(learning_rate=1e-4)  # optimizer
nfm = NormalizingFlowModel(nf, optimizer=opt, batch_size=ds.X.shape[0])
```

`int(1e2)` is an odd way of writing `100`!

KerasModel will create the Adam optimizer for you automatically.  You can just pass the argument `learning_rate=1e-4` to the model's constructor.

Are you really sure about `batch_size=ds.X.shape[0]`?  That's saying it shouldn't break the dataset into batches, and should use the entire dataset for every step.  That's a very unusual thing to do, so if that's really what you meant it needs explanation.

> Here's an example where the Tanimoto similarity scores are low! There are no molecules in QM9 that are very similar to our generated sample.

I don't think that's true.  Doesn't QM9 contain every possible molecule in this size range?  I think what you meant is that they aren't in the subset of 2000 molecules you used for training.

**Comment by peastman** [comment]
> Not too bad! The normalizing flow is a pretty good mapping between the multivariate Gaussian and the target distribution.

I don't think we can conclude that.  All you've done at this point is show that the loss decreased over training.  To determine whether it's actually a good mapping, you need to evaluate it on an independent validation set.

**Comment by ncfrey** [comment]
Thanks @peastman! 

> Looks good! I have a few suggestions.
> 
> A lot of readers probably aren't familiar with SELFIES. It would be helpful to add a few sentences discussing their advantages over SMILES. You also could show the two representations of a few molecules to help people understand the differences.
> 

Sounds good, will do.

> Rather than downloading files from AWS and loading them with pandas, why not just call `dc.molnet.load_qm9()`?
> 
Since I only need the SMILES string and I haven't written the SELFIES featurizer in DeepChem yet, I'm just pulling the files down from AWS.

> I don't understand what the permutation layer is for. It seems to just swap the first and second halves of the array. Can you add more explanation of that?
> 
Will do.
> This code
> 
> ```python
> for epoch in range(max_epochs):  # max_epochs
>     loss = nfm.fit(ds, nb_epoch=1)
>     losses.append(loss)
> ```
> 
> can be simplified to just
> 
> ```python
> loss = nfm.fit(ds, nb_epoch=1, all_losses=all_losses)
> ```
> 
+1
> A few points about the following code:
> 
> ```python
> max_epochs = int(1e2)  # maximum number of epochs of the training
> opt = Adam(learning_rate=1e-4)  # optimizer
> nfm = NormalizingFlowModel(nf, optimizer=opt, batch_size=ds.X.shape[0])
> ```
> 
> `int(1e2)` is an odd way of writing `100`!
> 
Yes haha, that's a holdover from when I was changing the epoch count often. I'll switch to `100`.
> KerasModel will create the Adam optimizer for you automatically. You can just pass the argument `learning_rate=1e-4` to the model's constructor.
> 
+1
> Are you really sure about `batch_size=ds.X.shape[0]`? That's saying it shouldn't break the dataset into batches, and should use the entire dataset for every step. That's a very unusual thing to do, so if that's really what you meant it needs explanation.
> 
I'll experiment with some more typical batch sizes and change that.
> > Here's an example where the Tanimoto similarity scores are low! There are no molecules in QM9 that are very similar to our generated sample.
> 
> I don't think that's true. Doesn't QM9 contain every possible molecule in this size range? I think what you meant is that they aren't in the subset of 2000 molecules you used for training.

I compared the generated molecules to the full QM9 dataset, so I think the generated molecules that aren't in QM9 are only because they imposed some stability criteria and charge neutrality. I'll look more into this.

**Comment by peastman** [comment]
> Since I only need the SMILES string and I haven't written the SELFIES featurizer in DeepChem yet, I'm just pulling the files down from AWS.

Using molnet will be both simpler and more familiar to readers.  The SMILES strings are in `dataset.ids`.

**Comment by seyonechithrananda** [comment]
This is really nice! On my end, I think it would be cool if at the end the tutorial could touch on some relevant work that is using normalizing flows for molecular generation since there are several papers on the molecular graph generation side. Aside from that, perhaps some background on SELFIES (you can just take the description from the SELFIES README) may be of value to users, to give a better understanding of why it makes sense to convert to SELFIES and not just one-hot encode SMILES. 

Also, perhaps a little bit of the math behind masked autoregressive flows could be interesting in the part where you introduce them! I believe we discussed some of the various papers on our first call! Overall, this is a really nice use of SELFIES, and I'm really happy this API + tutorial worked out!

**Papers of interest**, which apply flow-based generative models to molecular graph generation, could be interesting to link in the tutorial at the end for those interested:
[https://arxiv.org/abs/2006.10137](https://arxiv.org/abs/2006.10137), [https://arxiv.org/pdf/2001.09382.pdf](https://arxiv.org/pdf/2001.09382.pdf), [https://arxiv.org/pdf/1905.11600.pdf](https://arxiv.org/pdf/1905.11600.pdf)

**Comment by ncfrey** [comment]
> This is really nice! On my end, I think it would be cool if at the end the tutorial could touch on some relevant work that is using normalizing flows for molecular generation since there are several papers on the molecular graph generation side. Aside from that, perhaps some background on SELFIES (you can just take the description from the SELFIES README) may be of value to users, to give a better understanding of why it makes sense to convert to SELFIES and not just one-hot encode SMILES.
> 
Thanks @seyonechithrananda! That sounds good, I'll include more info on SELFIES and link to more reading and hopefully be able to link to DeepChem's flow-based molecule graph generative model soon.
> Also, perhaps a little bit of the math behind masked autoregressive flows could be interesting in the part where you introduce them! I believe we discussed some of the various papers on our first call! Overall, this is a really nice use of SELFIES, and I'm really happy this API + tutorial worked out!
+1

**Comment by ncfrey** [comment]
I've incorporated the changes suggested by @peastman 
A few new issues that would be great to get feedback on:
- I fixed a bug in the QM9 loader, but for some reason the featurization process is really slow, so it takes a long time to run `tasks, datasets, transformers = dc.molnet.load_qm9(featurizer='ECFP')` just to get the SMILES strings. I've tried different featurizers but no luck.
- I included `all_losses` as an argument to `fit`, but I also added a validation set and to get the val loss I'm calling the model's unique loss function on the validation set during each epoch. Any suggestions on consolidating this loop would be appreciated:
```
for epoch in range(max_epochs):
  loss = nfm.fit(train, nb_epoch=1, all_losses=losses)
  val_loss = nfm.create_nll(val.X)
  val_losses.append(val_loss.numpy())
```
- Usually somewhere between 30-40% of the generated molecules are valid. I thought this might be higher using the SELFIES representation, but maybe decoding back into SMILES is negating the advantage. @seyonechithrananda any thoughts?

**Comment by seyonechithrananda** [comment]
Hey @ncfrey, I believe I may know what may be causing lower validity % in the generated samples. In SELFIES, we define a generic set of semantic bond constraints, which define how many bonds each atom can take on. See here:

**Comment by seyonechithrananda** [comment]
Hey @ncfrey, I believe I may know what may be causing lower validity % in the generated samples. In SELFIES, we define a generic set of semantic bond constraints, which define how many bonds each atom can take on. See here: https://selfies-mirror.readthedocs.io/en/latest/selfies.html#advanced-functions

Essentially, after speaking with a colleague and taking a better look at the QM9 set used for distribution, you have to set the number of bonds of `N` to 5 to account for the nitro group. If you look at the `set_semantic_constraints()` method, you can define custom constraints like in the following example:

```
    constraints = sf.get_semantic_constraints()
    constraints['N'] = 5
    constraints['Br'] = 7
    constraints['Cl'] = 7
    constraints['I'] = 7
    sf.set_semantic_constraints(constraints)
```
Another likely issue is probably the constraint of charged atoms. QM9 includes [N+] and [O-] and so forth, and so you may want to constrain that as well with regards to the maximum number of bonds they can take on.

selfies.decoder() will then only generate SMILES that respect the bond constraints specified by the dictionary, adjusting the generated samples you get. If an atom or ion is not specified in bond_constraints, it will by default be constrained to 8 bonds. To change the default setting for unrecognized atoms or ions, set bond_constraints['?'] to the desired integer (between 1 and 8 inclusive).

Aside from the docs I sent, you may want to check out this example usage where we adjust the semantic bond constraints here: https://selfies-mirror.readthedocs.io/en/latest/selfies_examples.html#Advanced-Usage

I'm happy to try with helping adjust the constraints and see how it goes!

**Comment by seyonechithrananda** [comment]
@ncfrey Great to hear that the constraints helped boost validity to 86%! I think addressing this use case of SELFIES (especially for boosting performance on the various molecular task) will be a very important example for the tutorial! Feel free to link the documentation if needed for users in the tutorial.

**Comment by rbharath** [comment]
Are we good to merge in this PR? Excited to get the tutorial in :)

**Comment by ncfrey** [comment]
> Are we good to merge in this PR? Excited to get the tutorial in :)

@rbharath I am rerunning the notebook with the new constraints and boosted validity, and once I finish polishing it up I'll push changes and it should be ready to go!

> I think addressing this use case of SELFIES (especially for boosting performance on the various molecular task) will be a very important example for the tutorial!

@seyonechithrananda Definitely! I'm adding a bit of discussion on bond constraints and linking to the docs.

**Comment by ncfrey** [comment]
Alright should be good to go!

**Comment by seyonechithrananda** [comment]
Awesome!!!!

**Comment by MarioKrenn6240** [comment]
Hi @ncfrey , Seyone has just pointed me to this tutorial, great! We shall link it in the main SELFIES repo.

I have a question -- do i understand correctly that some SELFIES strings are invalid? If it's easily possible, could you please collect a few of them, and post them as issues in SELFIES? It should actually not happen -- so i expect either a bug or an unimplemented feature. Either would be useful to know.

Btw, we upload SELFIES1.0.2 to pip early next week including your functions (wanted to add another feature, but it is more involved, so we just go with this now.).

Thanks, Mario
